chain/jsonrpc-primitives/src/errors.rs

```diff
@@ -49,28 +72,86 @@ impl RpcError {
         };
         RpcError::new(-32_602, "Invalid params".to_owned(), Some(value))
     }
+
     /// Create a server error.
+    #[cfg(feature = "adversarial")]
     pub fn server_error<E: Serialize>(e: Option<E>) -> Self {
         RpcError::new(
             -32_000,
             "Server error".to_owned(),
             e.map(|v| to_value(v).expect("Must be representable in JSON")),
         )
     }
-    /// Create an invalid request error.
-    pub fn invalid_request() -> Self {
-        RpcError::new(-32_600, "Invalid request".to_owned(), None)
-    }
+
     /// Create a parse error.
     pub fn parse_error(e: String) -> Self {
-        RpcError::new(-32_700, "Parse error".to_owned(), Some(Value::String(e)))
+        RpcError {
+            code: -32_700,
+            message: "Parse error".to_owned(),
+            data: Some(Value::String(e.clone())),
+            error_struct: Some(RpcErrorKind::RequestValidationError(
+                RpcRequestValidationErrorKind::ParseError { error_message: e },
+            )),
+        }
     }
+
     pub fn serialization_error(e: String) -> Self {
-        RpcError::new(-32_000, "Server error".to_owned(), Some(Value::String(e)))
+        RpcError::new_internal_error(Some(Value::String(e.clone())), e)
+    }
+
+    /// Helper method to define extract INTERNAL_ERROR in separate RpcErrorKind
+    /// Returns HANDLER_ERROR if the error is not internal one
+    pub fn new_internal_or_handler_error(error_data: Option<Value>, error_struct: Value) -> Self {
+        if error_struct["name"] == "INTERNAL_ERROR" {
+            Self::new_internal_error(error_data, error_struct["info"]["error_message"].to_string())
```

Comment:
I believe `to_string` will produce unnecessary quotes like `"error_message": "\"some message\""`. Use `.get()` instead of `[]` and `as_str()` instead of `to_string()`, and handle potential None values and gracefully return a new InternalError with a message that InternalError happened during serializing InternalError :shrug: